File: tests/spread/integration/rustc/testfiles/test_std.rs
use std::fs::{self, File};
use std::io::{Read, Write};
use std::sync::{Arc, Mutex};
use std::thread;
use std::collections::HashMap;
use std::time::{Duration, Instant};

// each test returns Result<(), String> where Err contains error message
type TestResult = Result<(), String>;
type TestFunc = fn() -> TestResult;

fn main() {
    println!("Running basic standard library checks...");

    // array of function pointers, and names for reporting
    let tests: [(TestFunc, &str); 7] = [
        (check_collections, "Collections"),
        (check_arithmetic, "Arithmetic"),
        (check_file_io, "File I/O"),
        (check_concurrency, "Concurrency"),
        (check_mem, "Memory"),
        (check_time, "Time"),
        (check_env, "Environment Variables and Args"),
    ];

    let mut overall_result: bool = true;
    for (i, (test, name)) in tests.iter().enumerate() {
        let result = test();
        match result {
            Ok(_) => println!("Test {} ({}) passed", i + 1, name),
            Err(e) => {
                println!("Test {} ({}) failed: {}", i + 1, name, e);
                overall_result = false;
            }
        }
    }

    if !overall_result {
        println!("Some tests failed.");
        std::process::exit(1);
    } else {
        println!("\nAll basic checks passed!");
    }
}

fn assert(condition: bool, message: &str) -> TestResult {
    if condition {
        Ok(())
    } else {
        Err(message.to_string())
    }
}

////////////////////////////////////////////////////////////////////////////////

fn check_collections() -> TestResult {
    // Test String and Vec
    let mut s = String::from("hello");
    s.push_str(" world");
    assert(s == "hello world", "String manipulation failed")?;
    
    let mut v = Vec::new();
    v.push(1);
    v.push(2);
    v.push(3);
    assert(v == vec![1, 2, 3], "Vector manipulation failed")?;

    // Test HashMap
    let mut map = HashMap::new();
    map.insert("key1", 10);
    map.insert("key2", 20);
    assert(map.get("key1") == Some(&10), "HashMap get failed")?;
    assert(map.len() == 2, "HashMap length incorrect")?;
    map.remove("key1");
    assert(!map.contains_key("key1"), "HashMap remove failed")?;

    Ok(())
}

fn check_arithmetic() -> TestResult {
    // Test integer operations
    assert(10 + 5 == 15, "Integer addition failed")?;
    assert(10 - 5 == 5, "Integer subtraction failed")?;
    assert(10 * 5 == 50, "Integer multiplication failed")?;
    assert(10 / 5 == 2, "Integer division failed")?;
    assert(10 % 3 == 1, "Integer modulus failed")?;

    // Test floating point operations and edge cases
    let a = 10.0;
    let b = 5.0;
    assert(a / b == 2.0, "Floating point division failed")?;

    // Test division by zero results in infinity
    let c = 1.0;
    let d = 0.0;
    assert(c / d == f64::INFINITY, "Division by zero did not yield infinity")?;
    Ok(())
}

fn check_file_io() -> TestResult {
    let filename = "test_file.txt";
    let test_data = "Hello from Rust file I/O!";

    // Write to a file
    let mut file = File::create(filename).map_err(|e| format!("File creation failed: {}", e))?;
    file.write_all(test_data.as_bytes()).map_err(|e| format!("File write failed: {}", e))?;

    // Read from the file
    let mut file = File::open(filename).map_err(|e| format!("File open failed: {}", e))?;
    let mut contents = String::new();
    file.read_to_string(&mut contents).map_err(|e| format!("File read failed: {}", e))?;
    assert(contents == test_data, "File contents do not match")?;

    // Clean up
    fs::remove_file(filename).map_err(|e| format!("File removal failed: {}", e))?;
    Ok(())
}

fn check_concurrency() -> TestResult {
    let counter = Arc::new(Mutex::new(0));
    let mut handles = vec![];

    for _ in 0..10 {
        let counter_clone = Arc::clone(&counter);
        let handle = thread::spawn(move || {
            let mut num = counter_clone.lock().unwrap();
            *num += 1;
        });
        handles.push(handle);
    }

    for handle in handles {
        handle.join().map_err(|e| format!("Thread join failed: {:?}", e))?;
    }

    assert(*counter.lock().unwrap() == 10, "Counter value incorrect")?;
    Ok(())
}


fn check_mem() -> TestResult {
    // Test size_of for fundamental types
    assert(std::mem::size_of::<i32>() == 4, "size_of::<i32>() failed")?;
    assert(std::mem::size_of::<f64>() == 8, "size_of::<f64>() failed")?;
    
    // Test alignment of types
    assert(std::mem::align_of::<i32>() == 4, "align_of::<i32>() failed")?;
    assert(std::mem::align_of::<f64>() == 8, "align_of::<f64>() failed")?;

    // Test a struct with different alignments
    // spellchecker: ignore repr
    #[repr(C)]
    struct AlignedStruct {
        a: i8,
        b: i32,
    }
    assert(std::mem::size_of::<AlignedStruct>() == 8, "size_of::<AlignedStruct>() failed")?;
    
    Ok(())
}

fn check_time() -> TestResult {
    // Test high-resolution timer
    let start = Instant::now();
    let duration = Duration::from_millis(50);
    thread::sleep(duration);
    let elapsed = start.elapsed();
    assert(elapsed >= duration, &format!("Sleep duration incorrect. Expected at least {:?}, got {:?}", duration, elapsed))?;
    
    // Test system time
    let system_now = std::time::SystemTime::now();
    let since_epoch = system_now.duration_since(std::time::UNIX_EPOCH)
                                .map_err(|e| format!("System time duration_since failed: {}", e))?;
    assert(since_epoch.as_secs() > 1700000000, "System time is likely incorrect or too far in the past")?;

    Ok(())
}

fn check_env() -> TestResult {
    let test_key = "RUST_TEST_VAR";
    let test_val = "rust_test_value";

    // Set and get an environment variable
    std::env::set_var(test_key, test_val);
    let var_val = std::env::var(test_key).map_err(|e| format!("Failed to get env var: {}", e))?;
    assert(var_val == test_val, "Failed to get correct environment variable value")?;
    
    // Test that args() returns at least one element (the program name)
    let args: Vec<String> = std::env::args().collect();
    assert(args.len() >= 1, "Command line arguments check failed")?;
    
    Ok(())
}

File: tests/spread/integration/gcc-13-powerpc64le-linux-gnu/testfiles/test_std.c
#include "test_std.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <pthread.h>
#include <unistd.h>
#include <time.h>
#include <sys/time.h>
#include <errno.h>
#include <assert.h>
#include <stdint.h>
#include <limits.h>
#include <stdatomic.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <stdalign.h>
#include <signal.h>


int main() {
    printf("Running basic standard library checks...\n");
    TestCase tests[] = {
        {"Arithmetic", check_arithmetic},
        {"File I/O", check_file_io},
        {"Concurrency", check_concurrency},
        {"Memory", check_memory},
        {"Time", check_time},
        {"Environment Variables and Args", check_env},
        {"Atomics", check_atomic},
        {"Process Control", check_process_control},
        {"Signals", check_signals},
    };

    int overall_result = 1;
    for (int i = 0; i < sizeof(tests)/sizeof(tests[0]); i++) {
        TestResult result;
        tests[i].test_func(&result);
        if (result.passed) {
            printf("Test %d (%s) passed\n", i + 1, tests[i].name);
        } else {
            printf("Test %d (%s) failed: %s\n", i + 1, tests[i].name, result.message);
            overall_result = 0;
        }
    }

    if (!overall_result) {
        printf("Some tests failed.\n");
        return 1;
    } else {
        printf("\nAll basic checks passed!\n");
        return 0;
    }
}

void check_arithmetic(TestResult* result) {
    if (10 + 5 != 15) { result->passed = 0; result->message = "Integer addition failed"; return; }
    if (10 - 5 != 5) { result->passed = 0; result->message = "Integer subtraction failed"; return; }
    if (10 * 5 != 50) { result->passed = 0; result->message = "Integer multiplication failed"; return; }
    if (10 / 5 != 2) { result->passed = 0; result->message = "Integer division failed"; return; }
    if (10 % 3 != 1) { result->passed = 0; result->message = "Integer modulus failed"; return; }

    double a = 10.0;
    double b = 5.0;
    if (a / b != 2.0) { result->passed = 0; result->message = "Floating point division failed"; return; }

// we have infinity only in C23
#ifdef INFINITY
    double c = 1.0;
    double d = 0.0;
    if (!(c / d == INFINITY)) { result->passed = 0; result->message = "Division by zero did not yield infinity"; return; }
#endif

    result->passed = 1;
    result->message = "Arithmetic tests passed";
}

void check_file_io(TestResult* result) {
    const char* filename = "test_file.txt";
    const char* test_data = "Hello from C file I/O!";
    FILE* file = fopen(filename, "w");
    if (!file) { result->passed = 0; result->message = "File creation failed"; return; }
    if (fwrite(test_data, 1, strlen(test_data), file) != strlen(test_data)) {
        fclose(file);
        result->passed = 0; result->message = "File write failed"; return;
    }
    fclose(file);
    file = fopen(filename, "r");
    if (!file) { result->passed = 0; result->message = "File open failed"; return; }
    char buffer[256];
    size_t read_bytes = fread(buffer, 1, sizeof(buffer)-1, file);
    if (read_bytes <= 0) {
        fclose(file);
        result->passed = 0; result->message = "File read failed"; return;
    }
    buffer[read_bytes] = '\0';
    fclose(file);
    if (strcmp(buffer, test_data) != 0) { result->passed = 0; result->message = "File contents do not match"; return; }
    if (remove(filename) != 0) { result->passed = 0; result->message = "File removal failed"; return; }
    result->passed = 1;
    result->message = "File I/O tests passed";
}

void* thread_func(void* arg) {
    int* counter = (int*)arg;
    for (int i = 0; i < 100000; i++) {
        __atomic_fetch_add(counter, 1, __ATOMIC_SEQ_CST);
    }
    return NULL;
}

void check_concurrency(TestResult* result) {
    pthread_t threads[10];
    int counter = 0;
    for (int i = 0; i < 10; i++) {
        if (pthread_create(&threads[i], NULL, thread_func, &counter) != 0) {
            result->passed = 0; result->message = "Thread creation failed"; return;
        }
    }
    for (int i = 0; i < 10; i++) {
        if (pthread_join(threads[i], NULL) != 0) {
            result->passed = 0; result->message = "Thread join failed"; return;
        }
    }
    if (counter != 1000000) { result->passed = 0; result->message = "Counter value incorrect"; return; }
    result->passed = 1;
    result->message = "Concurrency tests passed";
}

void check_memory(TestResult* result) {
    if (sizeof(int32_t) != 4) { result->passed = 0; result->message = "size_of::<int32_t>() failed"; return; }
    if (sizeof(double) != 8) { result->passed = 0; result->message = "size_of::<double>() failed"; return; }
    if (alignof(int32_t) != 4) { result->passed = 0; result->message = "align_of::<int32_t>() failed"; return; }
    if (alignof(double) != 8) { result->passed = 0; result->message = "align_of::<double>() failed"; return; }

    struct AlignedStruct {
        char a;
        int32_t b;
    };
    if (sizeof(struct AlignedStruct) != 8) { result->passed = 0; result->message = "size_of::<AlignedStruct>() failed"; return; }

    result->passed = 1;
    result->message = "Memory tests passed";
}

void check_time(TestResult* result) {
    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC, &start);
    usleep(50000); // sleep for 50 milliseconds
    clock_gettime(CLOCK_MONOTONIC, &end);
    long elapsed_ms = (end.tv_sec - start.tv_sec) * 1000 + (end.tv_nsec - start.tv_nsec) / 1000000;
    if (elapsed_ms < 50) { result->passed = 0; result->message = "Sleep duration incorrect"; return; }

    struct timeval tv;
    gettimeofday(&tv, NULL);
    if (tv.tv_sec < 1700000000) { result->passed = 0; result->message = "System time is likely incorrect or too far in the past"; return; }

    result->passed = 1;
    result->message = "Time tests passed";
}

void check_env(TestResult* result) {
    const char* test_key = "C_TEST_VAR";
    const char* test_val = "c_test_value";
    if (setenv(test_key, test_val, 1) != 0) { result->passed = 0; result->message = "Failed to set env var"; return; }
    char* var_val = getenv(test_key);
    if (!var_val || strcmp(var_val, test_val) != 0) { result->passed = 0; result->message = "Failed to get correct environment variable value"; return; }

    int argc = 1;
    char* argv[] = {"program_name", NULL};
    if (argc < 1) { result->passed = 0; result->message = "Command line arguments check failed"; return; }

    result->passed = 1;
    result->message = "Environment tests passed";
}

void check_atomic(TestResult* result) {
    atomic_int atomic_counter = 0;
    atomic_fetch_add(&atomic_counter, 1);
    if (atomic_load(&atomic_counter) != 1) { result->passed = 0; result->message = "Atomic increment failed"; return; }
    atomic_fetch_sub(&atomic_counter, 1);
    if (atomic_load(&atomic_counter) != 0) { result->passed = 0; result->message = "Atomic decrement failed"; return; }
    result->passed = 1;
    result->message = "Atomic tests passed";
}

void check_process_control(TestResult* result) {
    pid_t pid = fork();
    if (pid < 0) { result->passed = 0; result->message = "Fork failed"; return; }
    if (pid == 0) {
        // Child process
        exit(42);
    } else {
        // Parent process
        int status;
        if (waitpid(pid, &status, 0) < 0) { result->passed = 0; result->message = "Waitpid failed"; return; }
        if (WIFEXITED(status)) {
            int exit_status = WEXITSTATUS(status);
            if (exit_status != 42) { result->passed = 0; result->message = "Child exit status incorrect"; return; }
        } else {
            result->passed = 0; result->message = "Child did not exit normally"; return;
        }
    }
    result->passed = 1;
    result->message = "Process control tests passed";
}

void check_signals(TestResult* result) {
    // Simple signal handling test
    struct sigaction sa;
    sa.sa_handler = SIG_IGN; // Ignore the signal
    sigemptyset(&sa.sa_mask);
    sa.sa_flags = 0;
    if (sigaction(SIGUSR1, &sa, NULL) != 0) { result->passed = 0; result->message = "Sigaction failed"; return; }

    // Send the signal to self
    if (raise(SIGUSR1) != 0) { result->passed = 0; result->message = "Raise signal failed"; return; }

    result->passed = 1;
    result->message = "Signal handling tests passed";
}
